### PR TITLE
[Refactor] remove `array.prototype.filter` dependency

### PR DESCRIPTION
**Pull request by TimothyGu**
It unnecessarily adds 37 kB to the bundle size.

See https://bundlephobia.com/result?p=array.prototype.filter@1.0.0

### CODE REVIEW

**Review by ljharb — CHANGES_REQUESTED** [review]
This dependency was quite intentionally added, and I don't intend to remove it.

Bundlephobia is inaccurate and spreads FUD - you can't measure the bundle size impact of a package without measuring it in the context of an entire app - one which is highly likely to already have these particular dependencies present.

If the size is indeed unnecessary, that can be addressed by optimizing `array.prototype.filter`.

**Comment by TimothyGu** [comment]
I can indeed measure it in my app, https://timothygu.me/urltester/. My bundle size increased from 412K to 480K due to the update of `available-typed-arrays` from 1.0.2 to 1.0.3.

```
❯ grep 'resolved .*available-typed-arrays' yarn.lock && make && du -h bundle.js
  resolved "https://registry.yarnpkg.com/available-typed-arrays/-/available-typed-arrays-1.0.2.tgz#6b098ca9d8039079ee3f77f7b783c4480ba513f5"
npx browserify exports.js -o bundle.js
416K    bundle.js

❯ yarn upgrade
[...]

❯ grep 'resolved .*available-typed-arrays' yarn.lock && make && du -h bundle.js
  resolved "https://registry.yarnpkg.com/available-typed-arrays/-/available-typed-arrays-1.0.3.tgz#fb7d02445bfedefad79fad1fe47931163a227198"
npx browserify exports.js -o bundle.js
480K    bundle.js
```

The increase of the figure is even higher than what Bundlephobia indicates as I do not minify my bundle. I can assure you that available-typed-arrays is the only package the `yarn upgrade` command upgraded, but if you'd like to confirm, I've attached the diff of the yarn.lock file:

```diff
diff --git a/js/yarn.lock b/js/yarn.lock
index 69ace63..a8fb325 100644
--- a/js/yarn.lock
+++ b/js/yarn.lock
@@ -29,10 +29,16 @@ acorn@^7.0.0:
   resolved "https://registry.yarnpkg.com/acorn/-/acorn-7.4.1.tgz#feaed255973d2e77555b83dbc08851a6c63520fa"
   integrity sha512-nQyp0o1/mNdbTO1PO6kHkwSrmgZ0MT/jCCpNiwbUjGoRN4dlBhqJtoQuCnEOKzgTVwg0ZWiCoQy6SxMebQVh8A==
 
-array-filter@^1.0.0:
+array.prototype.filter@^1.0.0:
   version "1.0.0"
-  resolved "https://registry.yarnpkg.com/array-filter/-/array-filter-1.0.0.tgz#baf79e62e6ef4c2a4c0b831232daffec251f9d83"
-  integrity sha1-uveeYubvTCpMC4MSMtr/7CUfnYM=
+  resolved "https://registry.yarnpkg.com/array.prototype.filter/-/array.prototype.filter-1.0.0.tgz#24d63e38983cdc6bf023a3c574b2f2a3f384c301"
+  integrity sha512-TfO1gz+tLm+Bswq0FBOXPqAchtCr2Rn48T8dLJoRFl8NoEosjZmzptmuo1X8aZBzZcqsR1W8U761tjACJtngTQ==
+  dependencies:
+    call-bind "^1.0.2"
+    define-properties "^1.1.3"
+    es-abstract "^1.18.0"
+    es-array-method-boxes-properly "^1.0.0"
+    is-string "^1.0.5"
 
 asn1.js@^5.2.0:
   version "5.4.1"
@@ -63,11 +69,11 @@ assert@^2.0.0:
     util "^0.12.0"
 
 available-typed-arrays@^1.0.2:
-  version "1.0.2"
-  resolved "https://registry.yarnpkg.com/available-typed-arrays/-/available-typed-arrays-1.0.2.tgz#6b098ca9d8039079ee3f77f7b783c4480ba513f5"
-  integrity sha512-XWX3OX8Onv97LMk/ftVyBibpGwY5a8SmuxZPzeOxqmuEqUCOM9ZE+uIaD1VNJ5QnvU2UQusvmKbuM1FR8QWGfQ==
+  version "1.0.3"
+  resolved "https://registry.yarnpkg.com/available-typed-arrays/-/available-typed-arrays-1.0.3.tgz#fb7d02445bfedefad79fad1fe47931163a227198"
+  integrity sha512-CuPhFULixV/d89POo1UG4GqGbR7dmrefY2ZdmsYakeR4gOSJXoF7tfeaiqMHGOMrlTiJoeEs87fpLsBYmE2BMw==
   dependencies:
-    array-filter "^1.0.0"
+    array.prototype.filter "^1.0.0"
 
 balanced-match@^1.0.0:
   version "1.0.2"
@@ -459,7 +465,7 @@ elliptic@^6.5.3:
     minimalistic-assert "^1.0.1"
     minimalistic-crypto-utils "^1.0.1"
 
-es-abstract@^1.18.0-next.1, es-abstract@^1.18.0-next.2:
+es-abstract@^1.18.0, es-abstract@^1.18.0-next.1, es-abstract@^1.18.0-next.2:
   version "1.18.0"
   resolved "https://registry.yarnpkg.com/es-abstract/-/es-abstract-1.18.0.tgz#ab80b359eecb7ede4c298000390bc5ac3ec7b5a4"
   integrity sha512-LJzK7MrQa8TS0ja2w3YNLzUgJCGPdPOV1yVvezjNnS89D+VR08+Szt2mz3YB2Dck/+w5tfIq/RoUAFqJJGM2yw==
@@ -481,6 +487,11 @@ es-abstract@^1.18.0-next.1, es-abstract@^1.18.0-next.2:
     string.prototype.trimstart "^1.0.4"
     unbox-primitive "^1.0.0"
 
+es-array-method-boxes-properly@^1.0.0:
+  version "1.0.0"
+  resolved "https://registry.yarnpkg.com/es-array-method-boxes-properly/-/es-array-method-boxes-properly-1.0.0.tgz#873f3e84418de4ee19c5be752990b2e44718d09e"
+  integrity sha512-wd6JXUmyHmt8T5a2xreUwKcGPq6f1f+WwIJkijUqiGcJz1qqnZgP6XIK+QyIWU5lT7imeNxUll48bziG+TSYcA==
+
 es-to-primitive@^1.2.1:
   version "1.2.1"
   resolved "https://registry.yarnpkg.com/es-to-primitive/-/es-to-primitive-1.2.1.tgz#e55cd4c9cdc188bcefb03b366c736323fc5c898a"
```

Sure, Bundlephobia gets it wrong if the heavyweight package (array.prototype.filter) is already imported by some other package in the dependency tree. But that's not the case here, and assuming it is without further context is quite presumptuous, I must say.

File: index.js [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
-		'Uint32Array',
-		'Uint8Array',
-		'Uint8ClampedArray'
-	], function (typedArray) {
-		return typeof global[typedArray] === 'function';
-	});
+	var out = [];
+	for (var i = 0; i < possibleNames.length; i++) {
+		if (typeof global[possibleNames[i]] === 'function') {
+			out.push(possibleNames[i]);
```
Comment (ljharb) [inline index.js]: this introduces a runtime dependency on `Array.prototype.push`, which is something I'd previously avoided, and i'd consider a bug.

**Review by ljharb — COMMENTED** [review]
Thanks, that's helpful context.

Most apps will be targeting a wide range of browsers, and will be using es-abstract transitively, which is the bulk of the dep weight. I don't think it's that presumptuous. Certainly that won't be true for every app, though.

File: index.js (hunk 1 above)
Comment (TimothyGu) [reply index.js]: Would `out[out.length] = possibleNames[i]` work?

**Review by ljharb — APPROVED** [review]
sigh, alrighty. turns out this is used by browserify's `util` shim, so its bundle size is significantly more important than in a typical package.

**Comment by ljharb** [comment]
v1.0.4 published.

**Pull request merged**